ssl/statem/statem_srvr.c

```diff
@@ -114,6 +114,17 @@ int ossl_statem_server_read_transition(SSL *s, int mt)
                 return 1;
             }
         }
+        if (mt == SSL3_MT_CLIENT_KEY_EXCHANGE && s->s3->tmp.cert_request
```

Comment:
Better yet, maybe also separate all that hairy logic (also the `else if` branch) into  a separate `client_certificate_expected` method.